### PR TITLE
Improve support for request integration parameters

### PR DESCRIPTION
**Pull request by calvernaz**
Improve support for request parameters.

Terraform Sample - https://github.com/localstack/localstack-terraform-samples/tree/master/apigateway-appsync-integration

Links to: https://github.com/localstack/localstack/pull/7729

### CODE REVIEW

File: localstack/services/apigateway/helpers.py [hunk 1]
```diff
@@ -141,6 +141,92 @@ def resolve_references(self) -> dict:
...
+
+class RequestParametersResolver:
+    """
+    Integration request data mapping expressions
+    https://docs.aws.amazon.com/apigateway/latest/developerguide/request-response-data-mappings.html
+    """
+
+    def resolve(self, context: ApiInvocationContext) -> IntegrationParameters:
+        """
+        Resolve integration request parameters from the integration request parameters
```
Comment (whummer) [inline localstack/services/apigateway/helpers.py:158]: nit: this docstring seems a bit redundant, maybe we could concretize it a bit more (no need to change now, could be done in one of the next iterations.. 👍 )

**Review by whummer — APPROVED** [review]
LGTM! 👍

File: localstack/services/apigateway/helpers.py (hunk 1 above)
Comment (calvernaz) [reply localstack/services/apigateway/helpers.py:158]: Agreed - let me fix that is something meaningful from AWS

File: tests/unit/test_apigateway.py [hunk 2]
```diff
@@ -513,3 +515,36 @@ def test_construct_invocation_event(self):
...
+
+class TestRequestParameterResolver:
+    def test_resolve_request_parameters(self):
+        integration: Dict[str, Any] = {
+            "requestParameters": {
+                "integration.request.path.pathParam": "method.request.path.id",
+                "integration.request.querystring.baz": "method.request.querystring.baz",
+                "integration.request.querystring.token": "method.request.header.Authorization",
+                "integration.request.querystring.env": "stageVariables.enviroment",
+                "integration.request.header.Content-Type": "'application/json'",
```
Comment (joe4dev) [inline tests/unit/test_apigateway.py:528]: Are the double quotes intentional `"'` ?

File: localstack/services/apigateway/invocations.py [hunk 3]
```diff
@@ -477,13 +478,6 @@ def invoke_rest_api_integration_backend(invocation_context: ApiInvocationContext
                 table.put_item(Item=event_data)
                 response = requests_response(event_data)
                 return response
-        if uri.startswith("arn:aws:apigateway:") and ".appsync-api:" in uri:
```
Comment (joe4dev) [inline localstack/services/apigateway/invocations.py:480]: Consider adding a comment that `invoke_rest_api_integration_backend` is patched in ext (before moving to better abstractions; e.g., strategy pattern with a plugin-based factory).
In the long term, we should think about how to handle conditional pro extensions as this keeps reoccurring.

Comment (joe4dev) [reply localstack/services/apigateway/invocations.py:480]: 👍 for the planned refactoring `# TODO: refactor this to have a class per integration type to make it easy to test the encapsulated logic`
This looks hard to test and maintain at first glance :)

File: localstack/services/apigateway/helpers.py [hunk 4]
```diff
@@ -141,6 +141,96 @@ def resolve_references(self) -> dict:
...
+        integrations_parameters = {}
+        for k, v in request_params.items():
+            if v.lower() in method_request_params:
+                integrations_parameters[k] = method_request_params[v.lower()]
+            else:
+                # static values
+                integrations_parameters[k] = v.replace("'", "")
+
+        # build the integration parameters dict and not mutate in-place
+        # ApiInvocationContext because.
```
Comment (joe4dev) [inline localstack/services/apigateway/helpers.py]: nit: "because" of what?

File: localstack/services/apigateway/integration.py [hunk 5]
```diff
@@ -374,9 +382,9 @@ def invoke(self, invocation_context: ApiInvocationContext):
                     parsed_result = json.loads(str(result or "{}"))
             parsed_result = common.json_safe(parsed_result)
             parsed_result = {} if parsed_result is None else parsed_result
-            response.status_code = 200
             response._content = parsed_result
 
+        response.status_code = 200
```
Comment (joe4dev) [inline localstack/services/apigateway/integration.py:387]: good refactorings along the way 👍

File: tests/unit/test_apigateway.py [hunk 6]
```diff
@@ -513,3 +515,36 @@ def test_construct_invocation_event(self):
                 t["is_base64_encoded"],
             )
             assert result == t["expected"]
+
+
+class TestRequestParameterResolver:
+    def test_resolve_request_parameters(self):
```
Comment (joe4dev) [inline tests/unit/test_apigateway.py:521]: I guess that's more of an internal test. Preferably, we have some snapshot tests covering the e2e API.

**Review by joe4dev — APPROVED** [review]
LGTM 👍 

just a few minor suggestions

File: tests/unit/test_apigateway.py (hunk 2 above)
Comment (calvernaz) [reply tests/unit/test_apigateway.py:528]: Yep, part of the spec

File: localstack/services/apigateway/invocations.py (hunk 3 above)
Comment (calvernaz) [reply localstack/services/apigateway/invocations.py:480]: This is actually how it is being implemented now. Each backend integration has a class with a set of steps that are always required, request parameters resolvers, request templates, response parameters resolvers, and response templates.

Happy to hear your thoughts!

File: tests/unit/test_apigateway.py (hunk 6 above)
Comment (calvernaz) [reply tests/unit/test_apigateway.py:521]: unit test, yea. There are integration tests that use will use this, pretty much any backend integration and that can be snapshotted

**Pull request merged**